Fixed #21221 -- Made forms Media use staticfiles to serve files if available.

codingjoe opened the pull request:
Previously form assets defined as a forms media have been served using a joined
absolute path consisting of the `STATIC_URL` and defined relative path.
This commit changed the `absolute_path` method to serve form media using
`django.contrib.staticfiles` if configured.

This will allow users to server their form media from a CDN or any other
staticfiles storage of their choosing.

collinanderson commented on django/forms/widgets.py:
```diff
@@ -9,7 +9,9 @@
 
+from django.apps import apps
 from django.conf import settings
+from django.contrib.staticfiles.storage import staticfiles_storage
```
In general django doesn't allow contrib imports from core, but it might be worth making an exception in this case.

codingjoe replied:
I know, I don't like it ether. But I don't have a better idea.

collinanderson commented:
Do you think this is a better approach than https://github.com/claudep/django/commit/2f5c6892523e80d57f9b50bb7bc2e5dcfc07426b ?

charettes commented:
I think I prefer Claude's approach since it fixes the issue at the `static` tag level and not only for `Media` objects.

In all cases the import from `django.contrib.staticfiles` should be conditional.

codingjoe commented:
@collinanderson  I agree, Claude's approach is cleaner, sadly I didn't know it existed.
But I changes the behaviour beyond the ticket's scope. That actually changes how the static template tag works.
I don't know if this is really intended.

codingjoe commented:
@charettes Claude's implementation is redundant with https://github.com/django/django/blob/master/django/contrib/admin/templatetags/admin_static.py.
I can change my patch to use claude's approach and have have of the template tags use the other.

What's the verdict?

collinanderson commented:
removing the special handling in admin_static makes sense to me.

claudep commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,17 @@ following requirements are met:
 * you've collected all your static files by using the
   :djadmin:`collectstatic` management command
 
+.. versionchanged:: 1.9
```
We are targeting 1.10.

codingjoe commented:
@collinanderson Ok, did that and included 2.0 deprecation for the obsolete admin template tag. Changed the `contrib.staticfiles` docs as well. Because you can now simply remain using `{% load static %}{% static "path/to/file" %}`

claudep commented:
Could you please add a release note?

claudep commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,17 @@ following requirements are met:
+    * you no longer have to use ``staticfiles`` :ttag:`static<staticfiles-static>`
+      template tag to refer to your static files in your templates.
+      The ``static`` template tag uses the ``staticfiles`` template tag,
+      if ``staticfiles`` is configured..
```
Double dots.

collinanderson commented on django/contrib/admin/templatetags/admin_static.py:
```diff
@@ -1,17 +1,17 @@
 
-_static = None
+warnings.warn('"admin_static" template tags will be'
+              ' removed in favour of "django.templatetags.static".',
```
"favor", we use US spellings.

codingjoe commented:
I added a note for both admin and staticfiles. I hope I caught it all.

collinanderson commented on docs/releases/1.10.txt:
```diff
@@ -35,6 +35,8 @@ Minor features
 * For sites running on a subpath, the default :attr:`URL for the "View site"
   link <django.contrib.admin.AdminSite.site_url>` at the top of each admin page
   will now point to ``request.META['SCRIPT_NAME']`` if set, instead of ``/``.
+* :mod:`admin_static <django.contrib.admin.templatetags.admin_static` is now
```
From jenkins:
/home/jenkins/workspace/docs/docs/releases/1.10.txt:38: WARNING: py:mod reference target not found: admin_static <django.contrib.admin.templatetags.admin_static

Missing a closing parens? (or maybe it's complaining that it's undocumented?)

collinanderson commented:
This all _seems_ ok to me. So far now one else has complained about the "import contrib from core".

jaylett commented on django/forms/widgets.py:
```diff
@@ -78,10 +79,24 @@ def render_css(self):
+
+        :param path: Relative or absolute path to static asset.
+        :type path: str
+        :param prefix: Added only if ``django.contrib.staticfiles`` is not enabled.
```
What's `prefix` actually used for?

jaylett replied:
I can't actually find a use for it in the source code, and it's not (AFAICT) a publicly documented API, so maybe we could deprecate it?

codingjoe replied:
Yes please, when should it be removed?

jaylett replied:
This would need approval by a core contributor, but as it's an undocumented API that probably isn't used it might be acceptable to just remove it and document the removal. If not, use a normal deprecation approach, which would mean marking as deprecated now and removing it I think at 2.0 (because it wasn't deprecated at 1.8 LTS, so it needs to still work with deprecation warning at 1.11 LTS).

jaylett replied:
Although ideally I'd like to know what it was intended for before recommending deprecating it.

jaylett commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,17 @@ following requirements are met:
+
+    * you no longer have to use ``staticfiles`` :ttag:`static<staticfiles-static>`
+      template tag to refer to your static files in your templates.
+      The ``static`` template tag uses the ``staticfiles`` template tag,
```
This seems slightly confusing to me (although it's the fault of history). We're talking specifically about the `static` template tag in the `static` template library, which will now use the `static` template tag in the `staticfiles` template library (if `staticfiles` is enabled).

This might be better phrased by focussing on the impact, ie that if `staticfiles` is enabled, the `static` template *library* will automatically use it, rather than trying to describe exactly what the change is?

codingjoe replied:
Good point, I can't find the proper reference in the current documentation tho.

jaylett replied:
Ah. That's…awkward.

I think that `{% load static %}` can be reasonably documented *here*, in the `staticfiles` documentation, even though it's not actually part of it; it would also be a good place to explain why this is the case (which is then a hint to 3PA authors, and could be cross-refed from the 3PA primer). Then a `versionchanged` tag to note that prior to <1.10> this didn't work like that.

However I note that the staticfiles documentation is…poor at the moment. So perhaps you could do something accurate but terse here, and then further work at some point can improve things. (Depends how much time you have right now, really.)

jaylett replied:
I was wrong; there are docs in [`ref/templates/builtins`](https://docs.djangoproject.com/en/1.8/ref/templates/builtins/#std:templatetag-static).

jaylett commented:
I believe this will also fix #18455.

jaylett commented:
We also need the `Form.Media` topic doc updating (`docs/topics/forms/media.txt), specifically the ["Paths in assets declarations"](https://docs.djangoproject.com/en/1.8/topics/forms/media/#paths-in-asset-definitions) section. That section is in any case quite outdated (it reads sensibly for the 1.3–1.4 period) and looks to me to be inaccurate (it mentions falling back to `MEDIA_URL`, which was removed in f59a0401e5d0e19568a038f2a50954d1a9a1526f — fortunately it looks like this is after the 1.9 branch, so we can just fix the documentation here, providing this gets accepted).

jaylett commented on django/templatetags/static.py:
```diff
@@ -150,5 +151,23 @@ def do_static(parser, token):
+    """
+    global _static
+    if _static is None:
+        if apps.is_installed('django.contrib.staticfiles'):
```
Wouldn't it be easier to put this check in `StaticNode.url()`? Also, you seem to have occluded the `do_static()`, which was previously the actual `static` template tag…this means that you aren't handling the `{% static PATH as VAR %}` case any more, I think.

timgraham commented:
buildbot, add to whitelist.

timgraham commented on django/contrib/admin/templatetags/admin_static.py:
```diff
@@ -1,17 +1,17 @@
 register = Library()
 
-_static = None
+warnings.warn('"admin_static" template tags will be'
```
I think you'll need to move this into the function itself, otherwise it looks like the warning will trigger even if you don't use the template tag (that seems to be the reason for the build failure).

codingjoe replied:
yes, saw that. I'll try to wrap it up later today.

codingjoe commented:
Ok, that got bigger than expected.
- I removed the undocumented and unused `prefix` argument.
- I moved the implementation into `StaticNode` and deprecated everything that now is obsolete.

timgraham commented on django/templatetags/static.py:
```diff
@@ -150,5 +156,14 @@ def do_static(parser, token):
+    """
+    Return ether a StaticNode or staticfiles storage url if configured.
+
+    :param path: relative path to static asset
```
We don't typically use this style of docstring. I prefer if you restructure it as a sentence.

codingjoe replied:
Well what style do we use? I couldn't any reference on how to style the signature documentation. I can do it the google way if you want me to.

timgraham replied:
Usually just sentences.

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -138,7 +147,10 @@ File Uploads
-* ...
+* Widget media gets now served from using ``staticfiles`` if
+  configured.
+* :meth:`django.forms.widgets.Media.absolute_path` removed
```
This would be under "Backwards incompatible changes" not "Minor Features" but there's no need to mention private API changes unless you find evidence they are widely used.

codingjoe replied:
No I searched on github, I could only fine one small repo that does monkey patching there.

timgraham commented on django/templatetags/static.py:
```diff
@@ -108,7 +109,12 @@ def render(self, context):
-        return urljoin(PrefixNode.handle_simple("STATIC_URL"), path)
+        if apps.is_installed('django.contrib.staticfiles'):
+            from django.contrib.staticfiles.storage import staticfiles_storage  # NoQA
+
```
no blank line needed

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2430,15 @@ slightly different call:
     <img src="{{ myphoto }}"></img>
 
-.. note::
+.. versionchanged:: 1.10
```
Keep in mind that these notes are removed in 2 two release, so they should be "self-contained" as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -35,6 +35,8 @@ Minor features
 * For sites running on a subpath, the default :attr:`URL for the "View site"
   link <django.contrib.admin.AdminSite.site_url>` at the top of each admin page
   will now point to ``request.META['SCRIPT_NAME']`` if set, instead of ``/``.
+* ``django.contrib.admin.templatetags.admin_static.static`` is now
```
Please move to "Features deprecated" section.

timgraham commented on tests/forms_tests/contrib/storages.py:
```diff
@@ -0,0 +1,8 @@
+# -*- coding: utf-8 -*-
```
"coding" is only required if the file has non-ASCII characters (it'll be removed when we drop Python 2 support too).

jaylett commented on django/templatetags/static.py:
```diff
@@ -150,5 +156,14 @@ def do_static(parser, token):
     return StaticNode.handle_token(parser, token)
 
 
+@register.simple_tag
```
This shouldn't be here; `static()` is a helper function (admittedly, I think, undocumented and I suspect there only for the benefit of admin's helper `admin_static` template library, although it may need a deprecation cycle to remove) and isn't a tag. You removed the static tag (the old `do_static()` function, decorated), which is needed.

codingjoe replied:
I moved it. Should be fine now.

jaylett replied:
Still seems to be there in the latest commit?

timgraham commented on django/contrib/staticfiles/templatetags/staticfiles.py:
```diff
@@ -1,16 +1,29 @@
 
 def static(path):
-    return staticfiles_storage.url(path)
+    # Implementation has been moved to django.templatetags.static.
```
It might be worth delaying this deprecation for a few releases, otherwise I fear it could become quite annoying if not impossible for third-party packages to adjust their templates to run deprecation warning free and support older versions of Django.

jaylett replied:
+1. You'll also actually get two warnings, won't you? One for the tag, one for the helper.

codingjoe replied:
Agreed,  but eventually, it should be deleted.
What do you want me to do?

mlorant replied:
Maybe it could be worth to introduce a new type of DeprecationWarning ? Something like `WillBeDeprecatedInFuture`, that would be visible only to Django developers (did not check if it was possible with the `warnings` module though)...

jaylett replied:
@codingjoe I think we probably want to leave a note in a comment (FIXME?) to deal with this later. It's a bit ugly, but introducing more deprecation warnings infrastructure feels overkill. We can retain the documentation that this is now deprecated, and introduce it as a warning later, at 1.11 or 2.0. @timgraham is that what you were thinking?

timgraham replied:
Right, I would create a "Someday/maybe" ticket and give it a keyword of "2.0" so we remember to start the deprecation then.

codingjoe replied:
Ok I removed the deprecation warning and added a comment. I'll create a new ticket once this is merged.

timgraham commented on django/contrib/admin/templatetags/admin_static.py:
```diff
@@ -1,17 +0,0 @@
-from django.apps import apps
```
I would put this on the same deprecation timeline as the staticfiles "static" tag. While it's not documented, I think many people overriding admin templates are using it, so I think it would cause a lot of pain for users trying to support multiple versions of Django if we removed it.

codingjoe replied:
done

timgraham commented on django/contrib/staticfiles/templatetags/staticfiles.py:
```diff
@@ -1,36 +1,19 @@
-        path = self.path.resolve(context)
-        return static(path)
+    # Implementation has been moved to django.templatetags.static.
+    # Deprecation should start in version 2.0
```
"version 2.0" -> "Django 2.0." (helps grep)

timgraham commented on django/forms/widgets.py:
```diff
@@ -77,12 +77,24 @@ def render_css(self):
+        Args:
+            path (str): Relative or absolute path to static asset.
+
+        Returns:
```
I don't find this style helpful given the return value is already mentioned in the first sentence. You can incorporate the arg into that sentence too: "Given a relative or absolute path to a static asset, return..."

codingjoe replied:
It would be helpful if django would commit to pep257 and numpy or google docstring format to shorten the review cycle.

timgraham commented on django/forms/widgets.py:
```diff
@@ -77,12 +77,24 @@ def render_css(self):
+
+        Returns:
+            str: Absolute path to static asset.
+
```
No blank line at the end of docstring please.

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,17 @@ following requirements are met:
 
+.. versionchanged:: 1.10
+
+    * you no longer have to use ``staticfiles`` :ttag:`static<staticfiles-static>`
```
You can drop the bullet and start it like this: "In older versions, you also had to use the..."

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,17 @@ following requirements are met:
+      template tag to refer to your static files in your templates.
+      The ``static`` template tag uses the ``staticfiles`` template tag,
+      if ``staticfiles`` is configured.
+
```
We always use 1 blank line between sections.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2430,16 @@ slightly different call:
-    :setting:`STATIC_URL` setting and the given path). Use that instead if you
-    have an advanced use case such as :ref:`using a cloud service to serve
-    static files<staticfiles-from-cdn>`::
+If :mod:`staticfiles<django.contrib.staticfiles>` contrib app is configured,
```
"If the..."
I see you just move this, but you can use "~django.contrib.staticfiles" instead of "staticfiles<~django.contrib.staticfiles>"

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2430,16 @@ slightly different call:
-    have an advanced use case such as :ref:`using a cloud service to serve
-    static files<staticfiles-from-cdn>`::
+If :mod:`staticfiles<django.contrib.staticfiles>` contrib app is configured,
+the static template tag, will serve all files from
```
premature line wrap
"from the storage defined in..."

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2430,16 @@ slightly different call:
 
-        {% load static from staticfiles %}
-        <img src="{% static "images/hi.jpg" %}" alt="Hi!" />
+    In the past static had to be loaded form :mod:`staticfiles<django.contrib.staticfiles>`
```
In older versions, ...
add backticks around static
~django.contrib.staticfiles

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -496,3 +500,12 @@ removed in Django 1.10 (please see the :ref:`deprecation timeline
 * The warning that :class:`~django.core.signing.Signer` issued when given an
   invalid separator is now a ``ValueError``.
+
+* ``django.forms.widgets.Media.absolute_path`` removed
```
We don't usually document private API removals unless we have evidence they are commonly used. I don't think that applies to these two items. (Also, the section you are in here is for features that went through the deprecation cycle.)

timgraham commented on django/contrib/admin/widgets.py:
```diff
@@ -6,12 +6,12 @@
 from django.forms.utils import flatatt
 from django.forms.widgets import RadioFieldRenderer
 from django.template.loader import render_to_string
+from django.templatetags.static import static
```
Shouldn't we be able to remove usage of `static()` in this file now that `absolute_path()` adds it?

timgraham commented on tests/forms_tests/tests/test_media.py:
```diff
@@ -520,3 +520,24 @@ def test_html_safe(self):
+@override_settings(
+    STATIC_URL='http://media.example.com/static/',
+    INSTALLED_APPS=('django.contrib.staticfiles', ),
+    STATICFILES_STORAGE='forms_tests.contrib.storages.StaticTestStorage'
```
We like to include a trailing comma in kwargs like this so if more items are added, we don't have to modify this line again.

timgraham commented on tests/forms_tests/contrib/storages.py:
```diff
@@ -0,0 +1,7 @@
+from django.contrib.staticfiles import storage
```
I think I would be more inclined to put these tests in `staticfiles_tests/test_forms.py` (new file).

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -138,7 +141,8 @@ File Uploads
 
-* ...
+* Widget media gets now served using ``staticfiles`` if
+  configured.
```
fits on previous line

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -98,7 +98,10 @@ Minor features
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-* ...
+* you no longer have to use ``staticfiles`` :ttag:`static<staticfiles-static>`
```
You (capitalize)
the staticfiles ...

timgraham commented:
I think we have two separate ideas here:
1. Make the builtin static tag use staticfiles if installed.
2. Use that tag in constructing the path to form media.
I don't know that it's worth splitting them into two commits at this point, but if we were starting over, that's what I would do.

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -138,7 +141,8 @@ File Uploads
 ^^^^^
 
-* ...
+* Widget media gets now served using ``staticfiles`` if
```
The Media class can be used on Forms too.

timgraham commented:
James' comment about updating `docs/topics/forms/media.txt` is pending too.

codingjoe commented:
Ok, I think I'm all done.

timgraham commented on docs/topics/forms/media.txt:
```diff
@@ -239,6 +240,14 @@ But if :setting:`STATIC_URL` is ``'http://static.example.com/'``::
     <script type="text/javascript" src="http://othersite.com/actions.js"></script>
 
 
+Or If staticfiles is configured using the `~django.contib.staticfiles.ManifestStaticFilesStorage`::
```
Or if ....
(wrap line at 79 characters too)

timgraham commented on tests/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
```
I think this file shouldn't be added (it's the reason for the test failures as far as I can tell).

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -205,18 +203,6 @@ def test_trailing_slash_required(self):
         response = self.client.get(add_url[:-1])
         self.assertRedirects(response, add_url, status_code=301)
 
-    def test_admin_static_template_tag(self):
```
Probably should keep this test until the deprecation completes?

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2430,16 @@ slightly different call:
 
-        {% load static from staticfiles %}
-        <img src="{% static "images/hi.jpg" %}" alt="Hi!" />
+    In older versions, static had to be loaded form :mod:`staticfiles<django.contrib.staticfiles>`
```
loaded from
```
:mod:`~django.contrib.staticfiles`
```

timgraham commented on docs/topics/forms/media.txt:
```diff
@@ -203,12 +203,13 @@ Paths in asset definitions
 path starts with ``/``, ``http://`` or ``https://``, it will be
 interpreted as an absolute path, and left as-is. All other paths will
-be prepended with the value of the appropriate prefix.
+be prepended with the value of the appropriate prefix. Is the
```
"If the..."
use
```
:mod:`staticfiles<django.contrib.staticfiles>`
```
as above.

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,16 @@ following requirements are met:
 
+.. versionchanged:: 1.10
+
+    In older versions, you also had to use the ``staticfiles`` :ttag:`static<staticfiles-static>`
```
wrap at 79 characters

timgraham commented on django/contrib/admin/helpers.py:
```diff
@@ -13,6 +12,7 @@
 from django.db.models.fields.related import ManyToManyRel
 from django.forms.utils import flatatt
 from django.template.defaultfilters import capfirst, linebreaksbr
+from django.templatetags.static import static
```
Looks like static usage can be removed in this file since it's used in Media.

timgraham commented on django/contrib/admin/options.py:
```diff
@@ -35,6 +34,7 @@
 from django.http import Http404, HttpResponseRedirect
 from django.http.response import HttpResponseBase
 from django.template.response import SimpleTemplateResponse, TemplateResponse
+from django.templatetags.static import static
```
Looks removable as above.

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,16 @@ following requirements are met:
+
+    In older versions, you also had to use the ``staticfiles`` :ttag:`static<staticfiles-static>`
+    template tag to refer to your static files in your templates.
+    The ``static`` template tag uses the ``staticfiles`` template tag,
```
now uses

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -285,11 +285,16 @@ following requirements are met:
+    In older versions, you also had to use the ``staticfiles`` :ttag:`static<staticfiles-static>`
+    template tag to refer to your static files in your templates.
+    The ``static`` template tag uses the ``staticfiles`` template tag,
+    if ``staticfiles`` is configured.
```
if django.contrib.staticfiles is installed.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2430,16 @@ slightly different call:
-    have an advanced use case such as :ref:`using a cloud service to serve
-    static files<staticfiles-from-cdn>`::
+If the :mod:`~django.contrib.staticfiles` contrib app is configured,
+the static template tag, will serve all files from
```
from the storage specified by STATICFILES_STORAGE

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -98,7 +98,10 @@ Minor features
+* You no longer have to use the ``staticfiles`` :ttag:`static<staticfiles-static>`
+  template tag to refer to your static files in your templates.
+  The ``static`` template tag uses the ``staticfiles`` template tag,
+  if ``staticfiles`` is configured.
```
... is installed (as above)

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -138,7 +141,8 @@ File Uploads
 ^^^^^
 
-* ...
+* Forms and widget media gets now served using ``staticfiles``
```
```
``Media``
```
(add backticks) 
"... are now served using django.contrib.staticfiles if installed."

timgraham commented:
There are a few conflicts to resolve and some new `admin_static` references to rename after the merge of d638cdc42acec608c1967f44af6be32a477c239f.

codingjoe commented:
@timgraham I rebased.

timgraham commented:
Please check my edits: http://dpaste.com/2WX4A62

codingjoe commented:
@timgraham I applied most of your edits. Not the once that changed the RST syntax tho. That seems a little out of scope. Also the edit on 1.4 release notes.

I put it in a separate commit to simplify the review I can squash later.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -2430,18 +2431,16 @@ slightly different call:
-        <img src="{% static "images/hi.jpg" %}" alt="Hi!" />
+    In older versions, you had to use ``{% load static from staticfiles %}`` in
+    your template to serve files from the storage defined in
+docs/releases/1.10.txt    :setting:`STATICFILES_STORAGE`. This is no longer required.
```
Edits for this file weren't applied correctly and the changes in the 1.4 release are needed to prevent a broken link.

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -333,40 +338,6 @@ simply specify a custom entry in the :setting:`CACHES` setting named
 
 .. currentmodule:: django.contrib.staticfiles.templatetags.staticfiles
```
This should be removed as it was in my patch. Did you apply the edits manually such that I should check them line by line?

codingjoe replied:
Yes I couldn't apply the diff :(

timgraham replied:
`patch -p1 -i raw.diff` should work.

codingjoe replied:
I applied it now and pushed again.

timgraham commented:
I've done those RST cleanups separately as requested: 890938e756247d7ea34f96cc0cc0f8fab0d62e9b.

codingjoe commented:
@timgraham I rebased and fixed the merging issues I found. Sorry that I didn't just apply your diff, but `fatal: corrupt patch at line 423`

timgraham commented:
merged in cf546e11ac76c8dec527e39ff8ce8249a195ab42, thanks!

jaylett commented:
Terrific work @codingjoe & @timgraham! So happy to see this land :-)

codingjoe commented:
Thank you so much @timgraham!

dsanders11 commented:
@codingjoe, @timgraham it may be a good idea to add one media path to `tests/staticfiles_tests/test_forms.py` that uses the `static` templatetag like admin widgets did before this change. I believe it works (seems multiple calls to static() return the correct path) but it was a concern of mine looking at this PR since I know I've sprinkled the use of the `static` templatetag in various `class Media` in my own project (and in other open source reusable apps) to ensure the correct URL is used, and it would be good to ensure that behavior remains unchanged after this change.

timgraham commented:
Sounds okay to me.

codingjoe commented:
@dsanders11 that is a very good point. It's not only Django but many third party apps, that used the `static` function to hack this feature. We should have a test to ensure nothing breaks.

I'm on vacation right now, but I'm sure I can spare a minute later this evening.
@timgraham how do I do that, can I just open a PR or do I have to create a new ticket?

timgraham commented:
You can send a pull request referencing the original ticket, "Refs #21221 -- ...".

codingjoe commented:
@timgraham @dsanders11 here you go #5982
